### DISCUSSION TITLE
Is' use client 'meaningless without interaction or special circumstances?

### DISCUSSION THREAD

**Discussion by heygsc** [vercel/next.js#68052] [Help] [answered]
### Summary

When we want to use hooks such as useState or onClick in NextJS, we must use 'use client'.

If there are no hooks or other interactions in our server components, and we add 'use client', does it mean that 'use client' is meaningless at this point? Will it still be rendered as RSC?

So 'use client' only checks when needed? Isn't it decisive?

**Comment by icyJoseph** [comment, marked as answer]
> If there are no hooks or other interactions in our server components, and we add 'use client', does it mean that 'use client' is meaningless at this point? Will it still be rendered as RSC?
> 
> So 'use client' only checks when needed? Isn't it decisive?

The key thing that `use client` does, is that it tells the bundler that the JavaScript contained in the module (JS file and all of its imports), has to be bundled to be shipped to the browser. 

This allows for interactivity, because the JavaScript shipped to the client, is then used to attach event listeners, capture closures, attach callbacks, etc..., everything that happens in the browser runtime context, if you will. 

The recommendation is to not put `use client` until it is necessary. A component is by default a server component, and it won't send its JavaScript to the client. And if a component with out `use client`, is imported by a `use client` component, then it is `opted-into` the client modules.

The great thing, is that, you can really have two trees overlapping each other, and through composition make them work together.

```js
// no use client annotation, server component by default, unless..
export const RenderPostViews = async ({postId}) => {
    const views = await database.getPost(postId);
    return <span>{views}</span>
}
```

```js
'use client'
// Some animation might use browser APIs, event listeners, etc, who knows
export const Animate =  ({children}) => {
    return <SomeAnimation>{children}</SomeAnimation>
}
```

And then have a page.js:

```js
// this would just be a server component by default
import { RenderPostViews } from './postviews';
import { Animate } from './animate';
 
export default function Page(params) {
    const postId = getPostIdeFromParams(params)
     return (
             <Animate>
                  <RenderPostViews  postId={postId} />
             </Animate>
     );
}
```

And now you have a server tree that renders stuff from a DB, weaved/mixed, together with a client tree that uses browsers APIs.

**Reply by heygsc** [reply]
> This allows for interactivity, because the JavaScript shipped to the client, is then used to attach event listeners, capture closures, attach callbacks, etc..., everything that happens in the browser runtime context, if you will.

Does it look like hydrate?
That means the server component definitely does not have a hydration process in the client browser, but 'use client'/client component means there is a possibility of hydration?

**Reply by icyJoseph** [reply]
The hydration step is a bit more complex that than. 

While a server component doesn't send their JavaScript to the client, it does, obviously, send its output. The hydration step compares the initial HTML tree, with the RSC tree it received from the server (in <script> tags with RSC strings), which includes Server and Client components. The order is important, because, otherwise, event listeners might malfunction, etc..

**Reply by heygsc** [reply]
> The hydration step is a bit more complex that than.
> 
> While a server component doesn't send their JavaScript to the client, it does, obviously, send its output. The hydration step compares the current HTML tree, with the tree it received from the server, which includes Server and Client components. The order is important, because, otherwise, event listeners might malfunction, etc..

Well, it's actually like this.

Can you take a look at a small question about another topic we talked about?
https://github.com/vercel/next.js/discussions/67878#discussioncomment-10121353

I feel like there isn't much official documentation about `React Server Component`, and there are many articles from other developers that are not detailed...
